### PR TITLE
Feature/datasetorder strict filters

### PR DESCRIPTION
**Pull request by abbesBenayache**

### CODE REVIEW

File: api/src/services/order.js [hunk 1]
```diff
@@ -105,11 +106,13 @@ const requiredRequesterOrAnyClause = getRequiredAddressOrAnyClause('requester');
...
 
-const getAddressOrAnyRestrictClause = (key) => (value) => {
+const getAddressOrAnyRestrictClause = (key) => (value, isStrict) => {
   const restrictKey = `order.${key}`;
   return (
     !isAny(value) && {
-      [restrictKey]: value ? { $in: [NULL_ADDRESS, value] } : NULL_ADDRESS,
+      [restrictKey]: value
+        ? { $in: isStrict ? [value] : [NULL_ADDRESS, value] }
+        : NULL_ADDRESS,
```
Comment (PierreJeanjacquot) [inline api/src/services/order.js:115]: we can do a little optimization for the DB here by not using `$in` if not needed

```suggestion
      [restrictKey]: value
        ? isStrict
          ? value
          : { $in: [NULL_ADDRESS, value] }
        : NULL_ADDRESS,
```

File: api/src/utils/order-utils.js [hunk 2]
```diff
@@ -120,6 +120,8 @@ const TAG_MAP = {
   gpu: 9,
 };
 
+const IS_STRICT_FILTER = false;
```
Comment (PierreJeanjacquot) [inline api/src/utils/order-utils.js:123]: not sure we want to store it in a constant, if we do I would rename it `IS_STRICT_FILTER_DEFAULT_VALUE`

File: api/test/api-common.test.js [hunk 3]
```diff
@@ -5554,6 +5608,151 @@ describe('API', () => {
...
+          expect(data.ok).toBe(true);
+          expect(data.count).toBe(ordersExcludingAnyApp.length);
+          expect(data.orders).toBeDefined();
+          expect(Array.isArray(data.orders)).toBe(true);
+          expect(data.orders.length).toBe(ordersExcludingAnyApp.length);
+          data.orders.forEach((e) => {
+            expect(
+              e.order.apprestrict === allowedApp ||
+                e.order.apprestrict !== utils.NULL_ADDRESS,
+            ).toBe(true);
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5694]: the check seems to be too permissive, we expect apprestrict to be allowedApp

```suggestion
            expect(e.order.apprestrict).toBe(allowedApp);
```

File: api/test/api-common.test.js [hunk 4]
```diff
@@ -5554,6 +5608,151 @@ describe('API', () => {
...
+          expect(data.ok).toBe(true);
+          expect(data.count).toBe(ordersExcludingAnyWorkerpool.length);
+          expect(data.orders).toBeDefined();
+          expect(Array.isArray(data.orders)).toBe(true);
+          expect(data.orders.length).toBe(ordersExcludingAnyWorkerpool.length);
+          data.orders.forEach((e) => {
+            expect(
+              e.order.workerpoolrestrict === allowedWorkerpool ||
+                e.order.workerpoolrestrict === utils.NULL_ADDRESS,
+            ).toBe(true);
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5723]: same

```suggestion
            expect(e.order.workerpoolrestrict).toBe(allowedWorkerpool);
```

File: api/test/api-common.test.js [hunk 5]
```diff
@@ -5554,6 +5608,151 @@ describe('API', () => {
...
+          expect(data.ok).toBe(true);
+          expect(data.count).toBe(ordersExcludingAnyRequester.length);
+          expect(data.orders).toBeDefined();
+          expect(Array.isArray(data.orders)).toBe(true);
+          expect(data.orders.length).toBe(ordersExcludingAnyRequester.length);
+          data.orders.forEach((e) => {
+            expect(
+              e.order.workerpoolrestrict === allowedWorkerpool ||
+                e.order.workerpoolrestrict === utils.NULL_ADDRESS,
+            ).toBe(true);
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5752]: same + bad copy/paste I guess workerpool => requester

```suggestion
            expect(e.order.requesterrestrict).toBe(allowedRequester);
```

File: api/test/api-common.test.js [hunk 6]
```diff
@@ -5554,6 +5608,151 @@ describe('API', () => {
...
+          expect(result.status).toBe(OK_STATUS);
+          expect(result.data.ok).toBe(true);
+          expect(result.data.count).toBe(publicOrders.length);
+          expect(result.data.orders).toBeDefined();
+          expect(Array.isArray(result.data.orders)).toBe(true);
+          expect(result.data.orders.length).toBe(20);
+          expect(result.data.nextPage).toBeDefined();
+        });
+
+        test('GET /datasetorders (app filter & isAppStrict): should exclude orders with "any" filtered app)', async () => {
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5668]:
```suggestion
        test('GET /datasetorders (app filter & isAppStrict): should exclude orders with "any" app authorized', async () => {
```

File: api/test/api-common.test.js [hunk 7]
```diff
@@ -5554,6 +5608,151 @@ describe('API', () => {
...
+          expect(data.orders.length).toBe(ordersExcludingAnyWorkerpool.length);
+          data.orders.forEach((e) => {
+            expect(
+              e.order.workerpoolrestrict === allowedWorkerpool ||
+                e.order.workerpoolrestrict === utils.NULL_ADDRESS,
+            ).toBe(true);
+          });
+        });
+
+        test('GET /datasetorders (requester filter & isRequesterStrict): should exclude orders with "any" filtered requester)', async () => {
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5727]:
```suggestion
        test('GET /datasetorders (requester filter & isRequesterStrict): should exclude orders with "any" requester authorized', async () => {
```

**Review by PierreJeanjacquot — CHANGES_REQUESTED** [review]
- rephrase openapi description
- small filter optimization
- some tests to fix

File: api/src/utils/order-utils.js (hunk 2 above)
Comment (abbesBenayache) [reply api/src/utils/order-utils.js:123]: commit 35d6cf28c5aa5a3517901bcb3f31bf64036c256b

File: api/test/api-common.test.js (hunk 7 above)
Comment (abbesBenayache) [reply api/test/api-common.test.js:5727]: commit : 6fab488418a00e8ef7004ae32cd22009218e443c

File: api/test/api-common.test.js [hunk 8]
```diff
@@ -5554,6 +5608,151 @@ describe('API', () => {
...
+          expect(data.orders.length).toBe(ordersExcludingAnyApp.length);
+          data.orders.forEach((e) => {
+            expect(
+              e.order.apprestrict === allowedApp ||
+                e.order.apprestrict !== utils.NULL_ADDRESS,
+            ).toBe(true);
+          });
+        });
+
+        test('GET /datasetorders (workerpool filter & isWorkerpoolStrict): should exclude orders with "any" filtered workerpool)', async () => {
```
[the remark this answers is not in the document]
Comment (abbesBenayache) [reply api/test/api-common.test.js:5698]: commit : 6fab488418a00e8ef7004ae32cd22009218e443c

File: api/test/api-common.test.js (hunk 6 above)
Comment (abbesBenayache) [reply api/test/api-common.test.js:5668]: commit : 6fab488418a00e8ef7004ae32cd22009218e443c

File: api/test/api-common.test.js (hunk 5 above)
Comment (abbesBenayache) [reply api/test/api-common.test.js:5752]: commit : 6fab488418a00e8ef7004ae32cd22009218e443c

File: api/test/api-common.test.js (hunk 4 above)
Comment (abbesBenayache) [reply api/test/api-common.test.js:5723]: commit : 6fab488418a00e8ef7004ae32cd22009218e443c

File: api/test/api-common.test.js (hunk 3 above)
Comment (abbesBenayache) [reply api/test/api-common.test.js:5694]: commit : 6fab488418a00e8ef7004ae32cd22009218e443c

File: api/src/services/order.js (hunk 1 above)
Comment (abbesBenayache) [reply api/src/services/order.js]: je propose de supprimer les `[expressions](https://eslint.org/docs/latest/rules/no-nested-ternary)` ternaires imbriquées :


const getAddressOrAnyRestrictClause = (key) => (value, isStrict) => {
  const restrictKey = `order.${key}`;
  if (isAny(value)) {
    return null;
  }
  const result = { [restrictKey]: NULL_ADDRESS };
  if (value) {
    result[restrictKey] = isStrict ? value : { $in: [NULL_ADDRESS, value] };
  }
  return result;
};

File: api/test/api-common.test.js [hunk 9]
```diff
@@ -5262,6 +5262,60 @@ describe('API', () => {
             });
         });
 
+        test('GET /datasetorders (invalid isAppStrict): should return validation error for invalid isAppStrict value', async () => {
+          const { data, status } = await request
+            .get(
+              buildQuery('/datasetorders', {
+                dataset: datasetAddress, // *
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5269]: add mandatory chainId
```suggestion
                chainId, // *
                dataset: datasetAddress, // *
```

File: api/test/api-common.test.js [hunk 10]
```diff
@@ -5262,6 +5262,60 @@ describe('API', () => {
...
+          );
+          expect(data.count).toBeUndefined();
+          expect(data.orders).toBeUndefined();
+        });
+
+        test('GET /datasetorders (invalid isWorkerpoolStrict): should return validation error for invalid isWorkerpoolStrict value', async () => {
+          const { data, status } = await request
+            .get(
+              buildQuery('/datasetorders', {
+                dataset: datasetAddress, // *
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5287]:
```suggestion
                chainId, // *
                dataset: datasetAddress, // *
```

File: api/test/api-common.test.js [hunk 11]
```diff
@@ -5262,6 +5262,60 @@ describe('API', () => {
...
+          );
+          expect(data.count).toBeUndefined();
+          expect(data.orders).toBeUndefined();
+        });
+
+        test('GET /datasetorders (invalid isRequesterStrict): should return validation error for invalid isRequesterStrict value', async () => {
+          const { data, status } = await request
+            .get(
+              buildQuery('/datasetorders', {
+                dataset: datasetAddress, // *
```
Comment (PierreJeanjacquot) [inline api/test/api-common.test.js:5305]:
```suggestion
                chainId, // *
                dataset: datasetAddress, // *
```

**Review by PierreJeanjacquot — APPROVED** [review]

**Pull request merged**